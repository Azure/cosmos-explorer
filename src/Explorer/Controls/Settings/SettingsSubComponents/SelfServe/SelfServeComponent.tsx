import React from "react";
import { Descriptor, InputType, SmartUiComponent } from "../../../SmartUi/SmartUiComponent";
import { SqlX } from "./SqlX";

export class SelfServeComponent extends React.Component {
  private onSubmit = async (currentValues: Map<string, InputType>): Promise<void> => {
    console.log(currentValues.get("instanceCount"), currentValues.get("instanceSize"));
  };

  private selfServeData: Descriptor = {
    onSubmit: this.onSubmit,
    root: {
      id: "root",
      info: {
        message: "Start at $24/mo per database",
        link: {
          href: "https://aka.ms/azure-cosmos-db-pricing",
          text: "More Details"
        }
      },
      children: [
        {
          id: "instanceCount",
          input: {
            label: "Instance Count",
            dataFieldName: "instanceCount",
            type: "number",
            min: 1,
            max: 5,
            step: 1,
            defaultValue: 1,
            inputType: "slider",
            onChange: (currentState: Map<string, InputType>, newValue: InputType): Map<string, InputType> => {
              currentState.set("instanceCount", newValue);
              if ((newValue as number) === 1) {
                currentState.set("instanceSize", "1Core4Gb");
              }
              return currentState;
            }
          }
        },
        {
          id: "instanceSize",
          input: {
            label: "Instance Size",
            dataFieldName: "instanceSize",
            type: "object",
            choices: [
              { label: "1Core4Gb", key: "1Core4Gb", value: "1Core4Gb" },
              { label: "2Core8Gb", key: "2Core8Gb", value: "2Core8Gb" },
              { label: "4Core16Gb", key: "4Core16Gb", value: "4Core16Gb" }
            ],
            defaultKey: "1Core4Gb",
            onChange: (currentState: Map<string, InputType>, newValue: InputType): Map<string, InputType> => {
              currentState.set("instanceSize", newValue);
              return currentState;
            }
          }
        }
      ]
    }
  };

  public render(): JSX.Element {
    //return <SmartUiComponent descriptor={this.selfServeData} />
    return <SmartUiComponent descriptor={SqlX.toSmartUiDescriptor()} />;
  }
}
