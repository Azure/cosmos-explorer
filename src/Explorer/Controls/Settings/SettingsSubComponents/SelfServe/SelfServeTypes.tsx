
import "reflect-metadata";
import { EnumItem, Info, InputTypeValue } from "../../../SmartUi/SmartUiComponent";

const modifyParentProperty = (children: {[key: string]: any}, parentProperty: string, property: string | symbol) : any => {
    if (parentProperty in children) {
        children[parentProperty][property] ={id: property, input: {}}
        return children
    } else {
        const keys = Object.keys(children)
        for(var i =0; i< keys.length; i++) {
            children[keys[i]] = modifyParentProperty(children[keys[i]], parentProperty, property)
            return children
        }
    }
    return children
}

export const InfoBar = (metadataKey: string, info: Info) => {
    return (target: any) => {
        let context = Reflect.getMetadata(metadataKey, target)
        if(!context) {
          context = {id: "root", info: info,  input: undefined, children: {} }
        } else {
          context.info = info
        }
        console.log("class context:" + JSON.stringify(context))
        Reflect.defineMetadata(metadataKey, context, target)  
    };
};

export const Property = (metadataKey: string, parentProperty?: string): PropertyDecorator => {
    return (target, property) => {
      let context = Reflect.getMetadata(metadataKey, target)
      if(!context) {
        context = {id: "root", info: undefined,  input: undefined, children: {} }
        context.children[property] = {id: property, input: {}}
      }
      if (parentProperty) {
          const prevContextValue  = JSON.stringify(context) 
          context.children = modifyParentProperty(context.children, parentProperty, property)
          if (JSON.stringify(context) === prevContextValue) {
              throw new Error(`${parentProperty} not defined. declare it before the child property with @Property decorator.`)
          }
      } else {
          context.children[property] = {id: property, input: {}}
      }
      console.log("props context:" + JSON.stringify(context))
      Reflect.defineMetadata(metadataKey, context, target)
    };
};

export const modifyInputTypes = (metadataKey: string, fieldName: string, value: any) : PropertyDecorator => {
    return (target, property) => {
        let context = Reflect.getMetadata(metadataKey, target)
        if(!context) {
            throw new Error("Incorrect order")
        }
        context.children[property].input[fieldName] = value
        //context = modifyType(property)
        console.log("props context:" + JSON.stringify(context))
        Reflect.defineMetadata(metadataKey, context, target)
      };
}

export const Type = (metadataKey: string, type: InputTypeValue): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "type", type)
};

export const Label = (metadataKey: string, label: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "label", label)
};

export const DataFieldName = (metadataKey: string, dataFieldName: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "dataFieldName", dataFieldName)
};

export const Min = (metadataKey: string, min: number): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "min", min)
};

export const Max = (metadataKey: string, max: number): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "max", max)
};

export const Step = (metadataKey: string, step: number): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "step", step)
};

export const DefaultValue = (metadataKey: string, defaultValue: any): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "defaultValue", defaultValue)
};

export const TrueLabel = (metadataKey: string, trueLabel: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "trueLabel", trueLabel)
};

export const FalseLabel = (metadataKey: string, falseLabel: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "falseLabel", falseLabel)
};

export const Choices = (metadataKey: string, choices: EnumItem[]): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "choices", choices)
};

export const DefaultKey = (metadataKey: string, defaultKey: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "defaultKey", defaultKey)
};

export const NumberInputType = (metadataKey: string, numberInputType: string): PropertyDecorator => {
    return modifyInputTypes(metadataKey, "inputType", numberInputType)
};
