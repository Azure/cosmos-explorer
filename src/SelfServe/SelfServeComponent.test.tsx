import React from "react";
import { shallow } from "enzyme";
import { SelfServeComponent, SelfServeComponentState } from "./SelfServeComponent";
import { NumberUiType, OnSaveResult, SelfServeDescriptor, SmartUiInput } from "./SelfServeTypes";

describe("SelfServeComponent", () => {
  const defaultValues = new Map<string, SmartUiInput>([
    ["throughput", { value: 450 }],
    ["analyticalStore", { value: false }],
    ["database", { value: "db2" }],
  ]);
  const updatedValues = new Map<string, SmartUiInput>([
    ["throughput", { value: 460 }],
    ["analyticalStore", { value: true }],
    ["database", { value: "db2" }],
  ]);

  const initializeMock = jest.fn(async () => new Map(defaultValues));
  const onSaveMock = jest.fn(async () => {
    return {
      operationStatusUrl: undefined,
      portalNotification: {
        titleTKey: "SampleMessageTitleKey",
        messageTKey: "SampleMessageTextKey",
      },
    } as OnSaveResult;
  });
  const refreshResult = {
    isUpdateInProgress: false,
    updateInProgressMessageTKey: "refresh performed successfully",
  };

  const onRefreshMock = jest.fn(async () => {
    return { ...refreshResult };
  });
  const onRefreshIsUpdatingMock = jest.fn(async () => {
    return { ...refreshResult, isUpdateInProgress: true };
  });

  const exampleData: SelfServeDescriptor = {
    initialize: initializeMock,
    onSave: onSaveMock,
    onRefresh: onRefreshMock,
    inputNames: ["throughput", "analyticalStore", "database"],
    root: {
      id: "root",
      info: {
        messageTKey: "Start at $24/mo per database",
        link: {
          href: "https://aka.ms/azure-cosmos-db-pricing",
          textTKey: "More Details",
        },
      },
      children: [
        {
          id: "throughput",
          input: {
            labelTKey: "Throughput (input)",
            dataFieldName: "throughput",
            type: "number",
            min: 400,
            max: 500,
            step: 10,
            defaultValue: 400,
            uiType: NumberUiType.Spinner,
          },
        },
        {
          id: "containerId",
          input: {
            labelTKey: "Container id",
            dataFieldName: "containerId",
            type: "string",
          },
        },
        {
          id: "analyticalStore",
          input: {
            labelTKey: "Analytical Store",
            trueLabelTKey: "Enabled",
            falseLabelTKey: "Disabled",
            defaultValue: true,
            dataFieldName: "analyticalStore",
            type: "boolean",
          },
        },
        {
          id: "database",
          input: {
            labelTKey: "Database",
            dataFieldName: "database",
            type: "object",
            choices: [
              { label: "Database 1", key: "db1" },
              { label: "Database 2", key: "db2" },
              { label: "Database 3", key: "db3" },
            ],
            defaultKey: "db2",
          },
        },
      ],
    },
  };

  const isEqual = (source: Map<string, SmartUiInput>, target: Map<string, SmartUiInput>): void => {
    expect(target.size).toEqual(source.size);
    for (const key of source.keys()) {
      expect(target.get(key)).toEqual(source.get(key));
    }
  };

  it("should render and honor save, discard, refresh actions", async () => {
    const wrapper = shallow(<SelfServeComponent descriptor={exampleData} />);
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();

    // initialize() and onRefresh() should be called and defaults should be set when component is mounted
    expect(initializeMock).toHaveBeenCalledTimes(1);
    expect(onRefreshMock).toHaveBeenCalledTimes(1);
    let state = wrapper.state() as SelfServeComponentState;
    isEqual(state.currentValues, defaultValues);

    // when currentValues and baselineValues differ, save and discard should not be disabled
    wrapper.setState({ currentValues: updatedValues });
    wrapper.update();
    state = wrapper.state() as SelfServeComponentState;
    isEqual(state.currentValues, updatedValues);
    const selfServeComponent = wrapper.instance() as SelfServeComponent;
    expect(selfServeComponent.isSaveButtonDisabled()).toBeFalsy();
    expect(selfServeComponent.isDiscardButtonDisabled()).toBeFalsy();

    // when errors exist, save is disabled but discard is enabled
    wrapper.setState({ hasErrors: true });
    wrapper.update();
    state = wrapper.state() as SelfServeComponentState;
    expect(selfServeComponent.isSaveButtonDisabled()).toBeTruthy();
    expect(selfServeComponent.isDiscardButtonDisabled()).toBeFalsy();

    // discard resets currentValues to baselineValues
    selfServeComponent.discard();
    state = wrapper.state() as SelfServeComponentState;
    isEqual(state.currentValues, defaultValues);
    isEqual(state.currentValues, state.baselineValues);

    // resetBaselineValues sets baselineValues to currentValues
    wrapper.setState({ baselineValues: updatedValues });
    wrapper.update();
    state = wrapper.state() as SelfServeComponentState;
    isEqual(state.baselineValues, updatedValues);
    selfServeComponent.updateBaselineValues();
    state = wrapper.state() as SelfServeComponentState;
    isEqual(state.baselineValues, defaultValues);
    isEqual(state.currentValues, state.baselineValues);

    // clicking refresh calls onRefresh.
    selfServeComponent.onRefreshClicked();
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(onRefreshMock).toHaveBeenCalledTimes(2);

    selfServeComponent.onSaveButtonClick();
    expect(onSaveMock).toHaveBeenCalledTimes(1);
  });

  it("getResolvedValue", async () => {
    const wrapper = shallow(<SelfServeComponent descriptor={exampleData} />);
    await new Promise((resolve) => setTimeout(resolve, 0));
    const selfServeComponent = wrapper.instance() as SelfServeComponent;

    const numberResult = 1;
    const numberPromise = async (): Promise<number> => {
      return numberResult;
    };
    expect(await selfServeComponent.getResolvedValue(numberResult)).toEqual(numberResult);
    expect(await selfServeComponent.getResolvedValue(numberPromise)).toEqual(numberResult);

    const stringResult = "result";
    const stringPromise = async (): Promise<string> => {
      return stringResult;
    };
    expect(await selfServeComponent.getResolvedValue(stringResult)).toEqual(stringResult);
    expect(await selfServeComponent.getResolvedValue(stringPromise)).toEqual(stringResult);
  });

  it("message bar and spinner snapshots", async () => {
    const newDescriptor = { ...exampleData, onRefresh: onRefreshIsUpdatingMock };
    let wrapper = shallow(<SelfServeComponent descriptor={newDescriptor} />);
    await new Promise((resolve) => setTimeout(resolve, 0));
    let selfServeComponent = wrapper.instance() as SelfServeComponent;
    selfServeComponent.onSaveButtonClick();
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();

    newDescriptor.onRefresh = onRefreshMock;
    wrapper = shallow(<SelfServeComponent descriptor={newDescriptor} />);
    await new Promise((resolve) => setTimeout(resolve, 0));
    selfServeComponent = wrapper.instance() as SelfServeComponent;
    selfServeComponent.onSaveButtonClick();
    await new Promise((resolve) => setTimeout(resolve, 0));
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ isInitializing: true });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ compileErrorMessage: "sample error message" });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
