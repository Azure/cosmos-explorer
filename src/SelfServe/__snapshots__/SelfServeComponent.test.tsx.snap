// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelfServeComponent message bar and spinner snapshots 1`] = `
<div
  style={
    Object {
      "overflowX": "auto",
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
        "padding": 10,
      }
    }
  >
    <StackItem>
      <StyledCommandBarBase
        items={
          Array [
            Object {
              "disabled": true,
              "iconProps": Object {
                "iconName": "Save",
              },
              "key": "save",
              "onClick": [Function],
              "text": "Save",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": true,
              "iconProps": Object {
                "iconName": "Undo",
              },
              "key": "discard",
              "onClick": [Function],
              "text": "Discard",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": false,
              "iconProps": Object {
                "iconName": "Refresh",
              },
              "key": "refresh",
              "onClick": [Function],
              "text": "Refresh",
            },
          ]
        }
        styles={
          Object {
            "root": Object {
              "paddingLeft": 0,
            },
          }
        }
      />
      <Separator
        styles={
          Object {
            "root": Object {
              "height": 1,
              "padding": 0,
              "selectors": Object {
                "::before": Object {
                  "background": undefined,
                },
              },
            },
          }
        }
      />
    </StackItem>
    <StyledMessageBarBase
      messageBarType={0}
    >
      <Text>
        refresh performed successfully
      </Text>
    </StyledMessageBarBase>
    <SmartUiComponent
      currentValues={
        Map {
          "throughput" => Object {
            "disabled": false,
            "hidden": false,
            "value": 450,
          },
          "analyticalStore" => Object {
            "disabled": false,
            "hidden": false,
            "value": false,
          },
          "database" => Object {
            "disabled": false,
            "hidden": false,
            "value": "db2",
          },
        }
      }
      descriptor={
        Object {
          "initialize": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "inputNames": Array [
            "throughput",
            "analyticalStore",
            "database",
          ],
          "onRefresh": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "onSave": [MockFunction] {
            "calls": Array [
              Array [
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
              ],
              Array [
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "root": Object {
            "children": Array [
              Object {
                "id": "throughput",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "throughput",
                  "defaultValue": 400,
                  "labelTKey": "Throughput (input)",
                  "max": 500,
                  "min": 400,
                  "placeholderTKey": undefined,
                  "step": 10,
                  "type": "number",
                  "uiType": "Spinner",
                },
              },
              Object {
                "id": "containerId",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "containerId",
                  "labelTKey": "Container id",
                  "placeholderTKey": undefined,
                  "type": "string",
                },
              },
              Object {
                "id": "analyticalStore",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "analyticalStore",
                  "defaultValue": true,
                  "falseLabelTKey": "Disabled",
                  "labelTKey": "Analytical Store",
                  "placeholderTKey": undefined,
                  "trueLabelTKey": "Enabled",
                  "type": "boolean",
                },
              },
              Object {
                "id": "database",
                "info": undefined,
                "input": Object {
                  "choices": Array [
                    Object {
                      "key": "db1",
                      "labelTKey": "Database 1",
                    },
                    Object {
                      "key": "db2",
                      "labelTKey": "Database 2",
                    },
                    Object {
                      "key": "db3",
                      "labelTKey": "Database 3",
                    },
                  ],
                  "dataFieldName": "database",
                  "defaultKey": "db2",
                  "labelTKey": "Database",
                  "placeholderTKey": undefined,
                  "type": "object",
                },
              },
            ],
            "id": "root",
            "info": Object {
              "link": Object {
                "href": "https://aka.ms/azure-cosmos-db-pricing",
                "textTKey": "More Details",
              },
              "messageTKey": "Start at $24/mo per database",
            },
          },
        }
      }
      disabled={true}
      getTranslation={[Function]}
      onError={[Function]}
      onInputChange={[Function]}
    />
  </Stack>
</div>
`;

exports[`SelfServeComponent message bar and spinner snapshots 2`] = `
<div
  style={
    Object {
      "overflowX": "auto",
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
        "padding": 10,
      }
    }
  >
    <StackItem>
      <StyledCommandBarBase
        items={
          Array [
            Object {
              "disabled": true,
              "iconProps": Object {
                "iconName": "Save",
              },
              "key": "save",
              "onClick": [Function],
              "text": "Save",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": true,
              "iconProps": Object {
                "iconName": "Undo",
              },
              "key": "discard",
              "onClick": [Function],
              "text": "Discard",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": false,
              "iconProps": Object {
                "iconName": "Refresh",
              },
              "key": "refresh",
              "onClick": [Function],
              "text": "Refresh",
            },
          ]
        }
        styles={
          Object {
            "root": Object {
              "paddingLeft": 0,
            },
          }
        }
      />
      <Separator
        styles={
          Object {
            "root": Object {
              "height": 1,
              "padding": 0,
              "selectors": Object {
                "::before": Object {
                  "background": undefined,
                },
              },
            },
          }
        }
      />
    </StackItem>
    <SmartUiComponent
      currentValues={
        Map {
          "throughput" => Object {
            "disabled": false,
            "hidden": false,
            "value": 450,
          },
          "analyticalStore" => Object {
            "disabled": false,
            "hidden": false,
            "value": false,
          },
          "database" => Object {
            "disabled": false,
            "hidden": false,
            "value": "db2",
          },
        }
      }
      descriptor={
        Object {
          "initialize": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "inputNames": Array [
            "throughput",
            "analyticalStore",
            "database",
          ],
          "onRefresh": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
              Array [],
              Array [],
              Array [],
              Array [],
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "onSave": [MockFunction] {
            "calls": Array [
              Array [
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
              ],
              Array [
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
              ],
              Array [
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
                Map {
                  "throughput" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": 450,
                  },
                  "analyticalStore" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": false,
                  },
                  "database" => Object {
                    "disabled": false,
                    "hidden": false,
                    "value": "db2",
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "root": Object {
            "children": Array [
              Object {
                "id": "throughput",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "throughput",
                  "defaultValue": 400,
                  "labelTKey": "Throughput (input)",
                  "max": 500,
                  "min": 400,
                  "placeholderTKey": undefined,
                  "step": 10,
                  "type": "number",
                  "uiType": "Spinner",
                },
              },
              Object {
                "id": "containerId",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "containerId",
                  "labelTKey": "Container id",
                  "placeholderTKey": undefined,
                  "type": "string",
                },
              },
              Object {
                "id": "analyticalStore",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "analyticalStore",
                  "defaultValue": true,
                  "falseLabelTKey": "Disabled",
                  "labelTKey": "Analytical Store",
                  "placeholderTKey": undefined,
                  "trueLabelTKey": "Enabled",
                  "type": "boolean",
                },
              },
              Object {
                "id": "database",
                "info": undefined,
                "input": Object {
                  "choices": Array [
                    Object {
                      "key": "db1",
                      "labelTKey": "Database 1",
                    },
                    Object {
                      "key": "db2",
                      "labelTKey": "Database 2",
                    },
                    Object {
                      "key": "db3",
                      "labelTKey": "Database 3",
                    },
                  ],
                  "dataFieldName": "database",
                  "defaultKey": "db2",
                  "labelTKey": "Database",
                  "placeholderTKey": undefined,
                  "type": "object",
                },
              },
            ],
            "id": "root",
            "info": Object {
              "link": Object {
                "href": "https://aka.ms/azure-cosmos-db-pricing",
                "textTKey": "More Details",
              },
              "messageTKey": "Start at $24/mo per database",
            },
          },
        }
      }
      disabled={false}
      getTranslation={[Function]}
      onError={[Function]}
      onInputChange={[Function]}
    />
  </Stack>
</div>
`;

exports[`SelfServeComponent message bar and spinner snapshots 3`] = `
<div
  style={
    Object {
      "overflowX": "auto",
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
        "padding": 10,
      }
    }
  >
    <StackItem>
      <StyledCommandBarBase
        items={
          Array [
            Object {
              "disabled": true,
              "iconProps": Object {
                "iconName": "Save",
              },
              "key": "save",
              "onClick": [Function],
              "text": "Save",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": true,
              "iconProps": Object {
                "iconName": "Undo",
              },
              "key": "discard",
              "onClick": [Function],
              "text": "Discard",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": true,
              "iconProps": Object {
                "iconName": "Refresh",
              },
              "key": "refresh",
              "onClick": [Function],
              "text": "Refresh",
            },
          ]
        }
        styles={
          Object {
            "root": Object {
              "paddingLeft": 0,
            },
          }
        }
      />
      <Separator
        styles={
          Object {
            "root": Object {
              "height": 1,
              "padding": 0,
              "selectors": Object {
                "::before": Object {
                  "background": undefined,
                },
              },
            },
          }
        }
      />
    </StackItem>
    <StyledSpinnerBase
      size={3}
    />
  </Stack>
</div>
`;

exports[`SelfServeComponent message bar and spinner snapshots 4`] = `
<StyledMessageBarBase
  messageBarType={1}
>
  <Text>
    sample error message
  </Text>
</StyledMessageBarBase>
`;

exports[`SelfServeComponent should render and honor save, discard, refresh actions 1`] = `
<div
  style={
    Object {
      "overflowX": "auto",
    }
  }
>
  <Stack
    tokens={
      Object {
        "childrenGap": 5,
        "padding": 10,
      }
    }
  >
    <StackItem>
      <StyledCommandBarBase
        items={
          Array [
            Object {
              "disabled": true,
              "iconProps": Object {
                "iconName": "Save",
              },
              "key": "save",
              "onClick": [Function],
              "text": "Save",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": true,
              "iconProps": Object {
                "iconName": "Undo",
              },
              "key": "discard",
              "onClick": [Function],
              "text": "Discard",
            },
            Object {
              "buttonStyles": Object {
                "root": Object {
                  "paddingLeft": 20,
                },
              },
              "disabled": false,
              "iconProps": Object {
                "iconName": "Refresh",
              },
              "key": "refresh",
              "onClick": [Function],
              "text": "Refresh",
            },
          ]
        }
        styles={
          Object {
            "root": Object {
              "paddingLeft": 0,
            },
          }
        }
      />
      <Separator
        styles={
          Object {
            "root": Object {
              "height": 1,
              "padding": 0,
              "selectors": Object {
                "::before": Object {
                  "background": undefined,
                },
              },
            },
          }
        }
      />
    </StackItem>
    <SmartUiComponent
      currentValues={
        Map {
          "throughput" => Object {
            "disabled": false,
            "hidden": false,
            "value": 450,
          },
          "analyticalStore" => Object {
            "disabled": false,
            "hidden": false,
            "value": false,
          },
          "database" => Object {
            "disabled": false,
            "hidden": false,
            "value": "db2",
          },
        }
      }
      descriptor={
        Object {
          "initialize": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "inputNames": Array [
            "throughput",
            "analyticalStore",
            "database",
          ],
          "onRefresh": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "onSave": [MockFunction],
          "root": Object {
            "children": Array [
              Object {
                "id": "throughput",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "throughput",
                  "defaultValue": 400,
                  "labelTKey": "Throughput (input)",
                  "max": 500,
                  "min": 400,
                  "placeholderTKey": undefined,
                  "step": 10,
                  "type": "number",
                  "uiType": "Spinner",
                },
              },
              Object {
                "id": "containerId",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "containerId",
                  "labelTKey": "Container id",
                  "placeholderTKey": undefined,
                  "type": "string",
                },
              },
              Object {
                "id": "analyticalStore",
                "info": undefined,
                "input": Object {
                  "dataFieldName": "analyticalStore",
                  "defaultValue": true,
                  "falseLabelTKey": "Disabled",
                  "labelTKey": "Analytical Store",
                  "placeholderTKey": undefined,
                  "trueLabelTKey": "Enabled",
                  "type": "boolean",
                },
              },
              Object {
                "id": "database",
                "info": undefined,
                "input": Object {
                  "choices": Array [
                    Object {
                      "key": "db1",
                      "labelTKey": "Database 1",
                    },
                    Object {
                      "key": "db2",
                      "labelTKey": "Database 2",
                    },
                    Object {
                      "key": "db3",
                      "labelTKey": "Database 3",
                    },
                  ],
                  "dataFieldName": "database",
                  "defaultKey": "db2",
                  "labelTKey": "Database",
                  "placeholderTKey": undefined,
                  "type": "object",
                },
              },
            ],
            "id": "root",
            "info": Object {
              "link": Object {
                "href": "https://aka.ms/azure-cosmos-db-pricing",
                "textTKey": "More Details",
              },
              "messageTKey": "Start at $24/mo per database",
            },
          },
        }
      }
      disabled={false}
      getTranslation={[Function]}
      onError={[Function]}
      onInputChange={[Function]}
    />
  </Stack>
</div>
`;
